src/qat/purr/compiler/builders.py

```diff
@@ -553,89 +596,111 @@ def measure(
-            # Find the pre-measure sync in the preceeding measure block and merge our
-            # values into it.
-            syncs[-1].add_channels(entangled_qubits)
-
-            # Find the post-measure sync, merge with ours, then remove it.
-            final_syncronize = syncs[0] + qubit
-            self._instructions.remove(syncs[0])
+        if full_measure_block and len(pre_syncs) >= 2 and len(pre_phase_resets) >= 1:
+            # Merge the first and last sync with the new.
+            pre_syncs[0] += new_syncs[0]
+            pre_syncs[-1] += new_syncs[-1]
+
             # reset all qubits
-            final_phase_reset = phase_resets[0] + entangled_qubits
-            self._instructions.remove(phase_resets[0])
+            pre_phase_resets[-1] += new_phase_resets[-1]
 
             # Add in our current changes.
-            self.add(
+            self.insert(
                 [
-                    MeasurePulse(measure_channel, **qubit.pulse_measure),
-                    acquire_instruction,
-                    final_syncronize,
-                    final_phase_reset,
-                ]
+                    val
+                    for val in new_measure_block
+                    if isinstance(val, (MeasurePulse, Acquire))
+                ],
+                index=self._instructions.index(pre_syncs[-1]),
             )
+        else:
+            self.add(new_measure_block)
+
+    def measure(
+        self,
+        qubit: Qubit,
+        axis: ProcessAxis = None,
+        output_variable: str = None,
+    ) -> "QuantumInstructionBuilder":
+        """
+        Adds a measure instruction. Important note: this only adds the instruction, not
+        any post-processing instructions as well. If you're wanting to perform generic
+        common operations use the more specific measurement types, as they add all the
+        additional information.
+        """
+        if axis is None:
+            axis = ProcessAxis.SEQUENCE
 
-            # Move all post-processing until after our newly-shifted syncronize block.
-            # Order matters here.
-            for pp in [
-                val
-                for val in reversed(previous_measure_block)
-                if isinstance(val, PostProcessing)
-            ]:
-                self._instructions.remove(pp)
-                self.add(pp)
+        if axis == ProcessAxis.SEQUENCE:
+            mode = AcquireMode.INTEGRATOR
+        elif axis == ProcessAxis.TIME:
+            mode = AcquireMode.SCOPE
         else:
-            self.add(
-                [
-                    Synchronize(entangled_qubits),
-                    MeasurePulse(measure_channel, **qubit.pulse_measure),
-                    acquire_instruction,
-                    Synchronize(qubit),
-                    PhaseReset(entangled_qubits),
-                ]
+            raise ValueError(f"Wrong measure axis '{str(axis)}'!")
+
+        entangled_qubits = list(self._entanglement_map.get(qubit, []))
```

Comment:
Out of scope and not feedback for this PR! But just want to note I think entanglement map should be removed. Say if you entangle 3 qubits and measure the first as a mid circuit measurement, you have to wait for the last one! But you might be doing some dynamic decoupling of that. This kind of analysis should happen later anyway!